File: dongle/ui.py
import os
from pathlib import Path


def run_picker(root, paths, is_workspace=False, cwd=None, initial_query=""):
    """Run the interactive TUI picker. All heavy imports are deferred to here."""
    # Lazy: prompt_toolkit is only imported when the picker actually opens
    from prompt_toolkit import Application
    from prompt_toolkit.keys import Keys
    from prompt_toolkit.layout import Layout, HSplit, Window
    from prompt_toolkit.layout.controls import FormattedTextControl
    from prompt_toolkit.key_binding import KeyBindings
    from prompt_toolkit.styles import Style

    from dongle.matcher import search
    from dongle.scanner import scan_paths, save_cache, find_project_root, CACHE_FILE
    from dongle.frecency import get_frecency_scores

    if paths is None:
        paths = scan_paths(root, is_workspace=is_workspace)

    frecency = get_frecency_scores()

    state = {
        "is_workspace": is_workspace,
        "paths": paths,
        "filtered": search(initial_query, paths, frecency),
        "index": 0,
        "query": initial_query,
        "chosen": None,
        "is_scanning": False,
    }

    kb = KeyBindings()

    @kb.add("c-c")
    @kb.add("escape")
    def _cancel(event):
        event.app.exit()

    @kb.add("enter")
    def _select(event):
        if state["filtered"]:
            state["chosen"] = state["filtered"][state["index"]]
        event.app.exit()

    @kb.add("up")
    @kb.add("c-p")
    def _up(event):
        if state["filtered"]:
            state["index"] = (state["index"] - 1) % len(state["filtered"])
        event.app.invalidate()

    @kb.add("down")
    @kb.add("c-n")
    def _down(event):
        if state["filtered"]:
            state["index"] = (state["index"] + 1) % len(state["filtered"])
        event.app.invalidate()

    @kb.add("c-u")
    def _clear(event):
        state["query"] = ""
        state["filtered"] = search("", state["paths"], frecency)
        state["index"] = 0
        event.app.invalidate()

    @kb.add("c-w")
    def _workspace(event):
        if not state["is_workspace"]:
            state["is_workspace"] = True
            state["is_scanning"] = True
            event.app.invalidate()
            ws_root = find_project_root(cwd or os.getcwd())
            state["paths"] = scan_paths(ws_root, is_workspace=True)
            state["is_scanning"] = False
            state["filtered"] = search(state["query"], state["paths"], frecency)
            state["index"] = 0
            event.app.invalidate()

    @kb.add("c-r")
    def _rescan(event):
        state["is_scanning"] = True
        event.app.invalidate()
        state["paths"] = scan_paths(root, is_workspace=state["is_workspace"])
        cache_file = Path.home() / ".dongle_workspace_cache.json" if state["is_workspace"] else CACHE_FILE
        cache_key = "WORKSPACE:" + root if state["is_workspace"] else root
        save_cache(cache_key, state["paths"], cache_file)
        state["is_scanning"] = False
        state["filtered"] = search(state["query"], state["paths"], frecency)
        state["index"] = 0
        event.app.invalidate()

    @kb.add(Keys.Any)
    def _type(event):
        if event.data and len(event.data) == 1 and event.data.isprintable():
            state["query"] += event.data
            # Optimization: When typing a new character, the result set can only shrink.
            # Instead of searching the full ~10,000 paths, we search the already-filtered subset.
            # This reduces CPU time per keystroke by ~90% when typing quickly.
            state["filtered"] = search(state["query"], state["filtered"], frecency)
            state["index"] = 0
            event.app.invalidate()

    @kb.add("backspace")
    def _backspace(event):
        state["query"] = state["query"][:-1]
        state["filtered"] = search(state["query"], state["paths"], frecency)
        state["index"] = 0
        event.app.invalidate()

    def get_prompt_text():
        mode = "Workspace" if state["is_workspace"] else "Project"
        display_root = root.replace(str(Path.home()), "~")
        n_filtered = len(state["filtered"])
        n_total = len(state["paths"])

        if state["is_scanning"]:
            status = "  [Scanning...]"
        elif state["query"]:
            status = f"  ({n_filtered}/{n_total})"
        else:
            status = f"  ({n_total} dirs)"

        return [
            ("class:title", f"  {mode}  "),
            ("class:title-path", display_root),
            ("class:title-count", f"{status}\n"),
            ("class:prompt", "  / "),
            ("class:query", state["query"]),
            ("class:cursor", "█"),  # block cursor
        ]

    def _highlight(display_p: str, query: str, selected: bool):
        sel = "class:selected" if selected else ""
        arrow = "❯ " if selected else "  "
        pad = "  "

        if query:
            q_lo = query.lower()
            p_lo = display_p.lower()
            idx = p_lo.find(q_lo)
            if idx >= 0:
                pre = display_p[:idx]
                mid = display_p[idx:idx + len(query)]
                post = display_p[idx + len(query):]
                hi = f"{sel},class:match" if selected else "class:match"
                return [
                    (sel, f"{pad}{arrow}{pre}"),
                    (hi, mid),
                    (sel, f"{post}\n"),
                ]

        return [(sel, f"{pad}{arrow}{display_p}\n")]

    def get_results_text():
        if not state["filtered"]:
            if state["query"]:
                return [
                    ("class:no-match", "    No matches.\n"),
                    ("class:hint", "    Ctrl+W — search across workspaces\n"),
                ]
            return [("class:hint", "    Start typing to filter...\n")]

        results = []
        start = max(0, state["index"] - 7)
        end = min(len(state["filtered"]), start + 15)

        for i in range(start, end):
            p = state["filtered"][i]
            display_p = p[0] if isinstance(p, (tuple, list)) else p
            results.extend(_highlight(display_p, state["query"], i == state["index"]))

        return results

    def get_footer_text():
        return [
            ("class:footer-key", " ↑↓"),
            ("class:footer", " nav  "),
            ("class:footer-key", "Enter"),
            ("class:footer", " select  "),
            ("class:footer-key", "^W"),
            ("class:footer", " workspace  "),
            ("class:footer-key", "^R"),
            ("class:footer", " rescan  "),
            ("class:footer-key", "^U"),
            ("class:footer", " clear  "),
            ("class:footer-key", "Esc"),
            ("class:footer", " cancel "),
        ]

    style = Style.from_dict({
        "title": "#5f9ea0 bold",
        "title-path": "#87ceeb",
        "title-count": "#666666",
        "prompt": "#00d787 bold",
        "query": "#ffffff bold",
        "cursor": "#00d787",
        "selected": "bg:#1a3a5c #ffffff bold",
        "match": "#ffaf00 bold",
        "selected,match": "bg:#1a3a5c #ffaf00 bold",
        "no-match": "#666666",
        "hint": "#444444",
        "footer": "#444444",
        "footer-key": "#777777 bold",
    })

    prompt_win = Window(
        content=FormattedTextControl(get_prompt_text, focusable=False),
        height=2,
    )
    results_win = Window(
        content=FormattedTextControl(get_results_text, focusable=False),
        height=15,
    )
    footer_win = Window(
        content=FormattedTextControl(get_footer_text, focusable=False),
        height=1,
    )

    layout = Layout(HSplit([prompt_win, results_win, footer_win]))

    app = Application(
        layout=layout,
        key_bindings=kb,
        style=style,
        full_screen=False,
        mouse_support=False,
    )

    app.run()
    return state["chosen"]

File: lib/src/ui.dart
import 'dart:io';
import 'dart:math';
import 'scanner.dart';
import 'frecency.dart';
import 'matcher.dart';

class PickerState {
  bool isWorkspace;
  List<dynamic> paths;
  List<dynamic> filtered;
  int index;
  String query;
  dynamic chosen;
  bool isScanning;

  PickerState({
    required this.isWorkspace,
    required this.paths,
    required this.filtered,
    this.index = 0,
    this.query = '',
    this.isScanning = false,
  });
}

class DonglePicker {
  final String root;
  final String? cwd;
  final Scanner scanner;
  final Frecency frecency;
  late Map<String, int> frecencyScores;

  DonglePicker({required this.root, this.cwd})
    : scanner = Scanner(),
      frecency = Frecency();

  dynamic run({
    bool isWorkspace = false,
    String initialQuery = '',
    List<dynamic>? preloadedPaths,
  }) {
    frecencyScores = frecency.getFrecencyScores();

    var paths =
        preloadedPaths ?? scanner.scanPaths(root, isWorkspace: isWorkspace);
    var filtered = search(initialQuery, paths, frecency: frecencyScores);

    var state = PickerState(
      isWorkspace: isWorkspace,
      paths: paths,
      filtered: filtered,
      query: initialQuery,
    );

    stdin.echoMode = false;
    stdin.lineMode = false;

    try {
      _render(state);

      while (true) {
        int byte = stdin.readByteSync();
        if (byte == -1) break;

        if (byte == 27) {
          // Escape sequence
          int next1 = stdin.readByteSync();
          if (next1 == 91) {
            // [
            int next2 = stdin.readByteSync();
            if (next2 == 65) {
              // Up
              _up(state);
            } else if (next2 == 66) {
              // Down
              _down(state);
            }
          } else {
            // Just Escape
            break;
          }
        } else if (byte == 3 || byte == 27) {
          // Ctrl+C or Escape
          break;
        } else if (byte == 13) {
          // Enter
          if (state.filtered.isNotEmpty) {
            state.chosen = state.filtered[state.index];
          }
          break;
        } else if (byte == 16) {
          // Ctrl+P (Up)
          _up(state);
        } else if (byte == 14) {
          // Ctrl+N (Down)
          _down(state);
        } else if (byte == 21) {
          // Ctrl+U
          state.query = '';
          state.filtered = search(
            state.query,
            state.paths,
            frecency: frecencyScores,
          );
          state.index = 0;
        } else if (byte == 23) {
          // Ctrl+W
          if (!state.isWorkspace) {
            state.isWorkspace = true;
            state.isScanning = true;
            _render(state);
            String wsRoot = scanner.findProjectRoot(
              cwd ?? Directory.current.path,
            );
            state.paths = scanner.scanPaths(wsRoot, isWorkspace: true);
            state.isScanning = false;
            state.filtered = search(
              state.query,
              state.paths,
              frecency: frecencyScores,
            );
            state.index = 0;
          }
        } else if (byte == 18) {
          // Ctrl+R
          state.isScanning = true;
          _render(state);
          state.paths = scanner.scanPaths(root, isWorkspace: state.isWorkspace);
          scanner.saveCache(
            state.isWorkspace ? "WORKSPACE:\$root" : root,
            state.paths,
          );
          state.isScanning = false;
          state.filtered = search(
            state.query,
            state.paths,
            frecency: frecencyScores,
          );
          state.index = 0;
        } else if (byte == 127 || byte == 8) {
          // Backspace
          if (state.query.isNotEmpty) {
            state.query = state.query.substring(0, state.query.length - 1);
            state.filtered = search(
              state.query,
              state.paths,
              frecency: frecencyScores,
            );
            state.index = 0;
          }
        } else {
          // Printable characters
          if (byte >= 32 && byte <= 126) {
            String c = String.fromCharCode(byte);
            String oldQuery = state.query;
            state.query += c;

            List<dynamic> pathsToSearch = state.paths;
            if (oldQuery.isNotEmpty && state.filtered.length < 50) {
              pathsToSearch = state.filtered;
            }

            state.filtered = search(
              state.query,
              pathsToSearch,
              frecency: frecencyScores,
            );
            state.index = 0;
          }
        }

        _render(state);
      }
    } finally {
      stdin.echoMode = true;
      stdin.lineMode = true;
      stdout.write('\x1B[18B');
      stdout.write('\x1B[J');
      stdout.write('\r');
    }

    return state.chosen;
  }

  void _up(PickerState state) {
    if (state.filtered.isNotEmpty) {
      state.index = (state.index - 1) % state.filtered.length;
      if (state.index < 0) state.index += state.filtered.length;
    }
  }

  void _down(PickerState state) {
    if (state.filtered.isNotEmpty) {
      state.index = (state.index + 1) % state.filtered.length;
    }
  }

  void _render(PickerState state) {
    stdout.write('\r');

    String mode = state.isWorkspace ? "Workspace" : "Project";
    String home = Platform.isWindows
        ? Platform.environment['USERPROFILE']!
        : Platform.environment['HOME']!;
    String displayRoot = root.replaceFirst(home, '~');
    int nFiltered = state.filtered.length;
    int nTotal = state.paths.length;

    String status = state.isScanning
        ? "  [Scanning...]"
        : (state.query.isNotEmpty
              ? "  (\$nFiltered/\$nTotal)"
              : "  (\$nTotal dirs)");

    stdout.write('\x1B[1m\x1B[38;5;73m  \$mode  \x1B[0m');
    stdout.write('\x1B[38;5;117m\$displayRoot\x1B[0m');
    stdout.write('\x1B[38;5;242m\$status\x1B[0m\n\r');

    stdout.write('\x1B[1m\x1B[38;5;42m  / \x1B[0m');
    stdout.write('\x1B[1m\x1B[37m\${state.query}\x1B[0m');
    stdout.write('\x1B[38;5;42m█\x1B[0m\x1B[K\n\r');

    int start = max(0, state.index - 7);
    int end = min(state.filtered.length, start + 15);

    int linesRendered = 0;

    if (state.filtered.isEmpty) {
      if (state.query.isNotEmpty) {
        stdout.write('\x1B[38;5;242m    No matches.\x1B[0m\x1B[K\n\r');
        stdout.write(
          '\x1B[38;5;238m    Ctrl+W — search across workspaces\x1B[0m\x1B[K\n\r',
        );
        linesRendered += 2;
      } else {
        stdout.write(
          '\x1B[38;5;238m    Start typing to filter...\x1B[0m\x1B[K\n\r',
        );
        linesRendered += 1;
      }
    } else {
      for (int i = start; i < end; i++) {
        var p = state.filtered[i];
        String displayP = (p is List) ? p[0] as String : p as String;
        bool selected = (i == state.index);

        _renderResultLine(displayP, state.query, selected);
        linesRendered++;
      }
    }

    for (int i = linesRendered; i < 15; i++) {
      stdout.write('\x1B[K\n\r');
    }

    stdout.write(
      '\x1B[38;5;240m─────────────────────────────────────────────────────────\x1B[0m\n\r',
    );
    stdout.write(
      '\x1B[1m\x1B[38;5;243m ↑↓\x1B[0m\x1B[38;5;238m nav  \x1B[0m'
      '\x1B[1m\x1B[38;5;243mEnter\x1B[0m\x1B[38;5;238m select  \x1B[0m'
      '\x1B[1m\x1B[38;5;243m^W\x1B[0m\x1B[38;5;238m workspace  \x1B[0m'
      '\x1B[1m\x1B[38;5;243m^R\x1B[0m\x1B[38;5;238m rescan  \x1B[0m'
      '\x1B[1m\x1B[38;5;243m^U\x1B[0m\x1B[38;5;238m clear  \x1B[0m'
      '\x1B[1m\x1B[38;5;243mEsc\x1B[0m\x1B[38;5;238m cancel \x1B[0m\x1B[K',
    );

    stdout.write('\r\x1B[18A');
  }

  void _renderResultLine(String displayP, String query, bool selected) {
    String selBg = selected ? '\x1B[48;5;17m\x1B[1m\x1B[37m' : '\x1B[0m';
    String arrow = selected ? '❯ ' : '  ';
    String pad = '  ';

    if (query.isNotEmpty) {
      String qLo = query.toLowerCase();
      String pLo = displayP.toLowerCase();
      int idx = pLo.indexOf(qLo);

      if (idx >= 0) {
        String pre = displayP.substring(0, idx);
        String mid = displayP.substring(idx, idx + query.length);
        String post = displayP.substring(idx + query.length);

        String hi = selected
            ? '\x1B[48;5;17m\x1B[1m\x1B[38;5;214m'
            : '\x1B[1m\x1B[38;5;214m';

        stdout.write(
          '\$selBg\$pad\$arrow\$pre\$hi\$mid\$selBg\$post\x1B[0m\x1B[K\n\r',
        );
        return;
      }
    }

    stdout.write('\$selBg\$pad\$arrow\$displayP\x1B[0m\x1B[K\n\r');
  }
}
